### ISSUE TITLE
call_contract_syscall causes a call to revert on Starknet but not in local

### ISSUE THREAD

**Issue by Th0rgal** [starkware-libs/cairo#5448]
# Bug Report

tldr: call_contract_syscall causes a call to revert on Starknet but it works fine in local (cairo-test).

**Cairo version:**
```
scarb 2.6.4 (c4c7c0bac 2024-03-19)
cairo: 2.6.3 (https://crates.io/crates/cairo-lang-compiler/2.6.3)
sierra: 1.5.0
```

**Context:**
I am building a short but quite logic intense Starknet contract allowing abstract interactions between calls. Instead of just being able to chain calls like in a normal multicall, you can take the output a previous call and use it as the output of a future call. I added some conditional executions feature (like skipping a call given the result of a previous call) and I am now trying to add a kind of "trycatch" feature (if call A fails, perform B, otherwise perform C, then perform D). To do so I need to be able to call a contract function and continue the execution even though the returned object is a Result::Err.

**Bug:**
The relevant code part of my multicall looks like this:
```rs
                    let call_result = call_contract_syscall(
                        build_input(@results, call.to).try_into().unwrap(),
                        build_input(@results, call.selector),
                        build_inputs(@results, call.calldata.span()).span()
                    );
                    results.append(call_result);
```
It then returns the list of results. The whole code is here: https://github.com/starknet-id/composable_multicall/blob/b01ef532a2b90064e1fe30a1603d3fd39db90e2f/src/contract.cairo#L176-L181

To test it I created a DummyContract with a simple assertion function (it takes two params, if they're not equal it panics).
```cairo
        // selector: 0x005562b3e932b4d139366854d5a2e578382e6a3b6572ac9943d55e7efbe43d00
        fn assert_eq(self: @ContractState, x: felt252, y: felt252) {
            assert(x == y, 'y does not equal y');
        }
```

I am using `scarb test` to simulate calls in local. These calls don't fail, instead they output me an array of results (which could include Errors, as expected). In this example I simulate:
1) assert_eq(1, 0) # should fail
2) if 1) failed, call dummy.one()
3) if 1) succeeded, call dummy.zero()
Which outputs:
```
[ Result::err('x does not equal y'), Result:ok([ 1 ]), Result::Err( 'skipped call' ) ]
```
The idea in a real use case would actually be to use `aggregate` to filter errors out so you get:
```
[ [1] ]
```
Relevant part of the test:
```cairo
    let catch_result = multicall
        .raw_aggregate(
            array![
                DynamicCall {
                    execution: Execution::Static,
                    to: DynamicFelt::Hardcoded(dummy.contract_address.into()),
                    selector: DynamicFelt::Hardcoded(EQU_SELECTOR),
                    calldata: array![DynamicCalldata::Hardcoded(1), DynamicCalldata::Hardcoded(0)]
                },
                DynamicCall {
                    execution: Execution::Catch(0), // we catch error on call 0
                    to: DynamicFelt::Hardcoded(dummy.contract_address.into()),
                    selector: DynamicFelt::Hardcoded(ONE_SELECTOR),
                    calldata: array![]
                },
                DynamicCall {
                    execution: Execution::Then(0), // we catch success on call 0
                    to: DynamicFelt::Hardcoded(dummy.contract_address.into()),
                    selector: DynamicFelt::Hardcoded(ZER_SELECTOR),
                    calldata: array![]
                },
            ]
        );
    assert(catch_result.len() == 3, 'wrong result length');
    assert(catch_result.at(0).is_err(), 'first call must be an error');
    match catch_result.at(1) {
        Result::Ok(_second_call) => {
            let second_call = *_second_call;
            assert(*second_call.at(0) == 1, 'wrong call result');
        },
        Result::Err(_) => { panic_with_felt252('Catch should be triggered'); }
    }
    match catch_result.at(2) {
        Result::Ok(_) => { panic_with_felt252('Then should not be triggered'); },
        Result::Err(err) => { assert(*err.at(0) == 'starknetid/call-skipped', 'wrong skip error'); }
    }
```
Whole test if you want to see extra examples:
https://github.com/starknet-id/composable_multicall/blob/b01ef532a2b90064e1fe30a1603d3fd39db90e2f/src/tests.cairo#L332-L432

**To reproduce:**
The call not failing in local (expected):
```
git clone https://github.com/starknet-id/composable_multicall/
cd composable_multicall
git checkout `b01ef532a2b90064e1fe30a1603d3fd39db90e2f`
scarb test -f test_catch_then_execution
```
The call failing on Starknet (unexpected):
- network: sepolia
- composable_multicall: `0x07a9013697371ce40d0306b4c810c6a4db9bfda119dd9ae1e8701c8e288d734b`
- dummy_contract: `0x001f40f0b0ad4322e13a35fae5400be2a2927a9b1ec7a1748158c1c456c49e56`

A call to multicalls.raw_aggregate( [ dummy_contract.assert_eq(1, 1) ] ) not failing (expected):
```json
curl --location 'https://free-rpc.nethermind.io/sepolia-juno/' \
--header 'Content-Type: application/json' \
--data '{
    "jsonrpc": "2.0",
    "method": "starknet_call",
    "params": {
        "request": {
            "contract_address": "0x07a9013697371ce40d0306b4c810c6a4db9bfda119dd9ae1e8701c8e288d734b",
            "entry_point_selector": "0x253d5ae28eea16e97fafcf8c5b2143ef58147be2687b9f15d9ec97992365fb8",
            "calldata": [
                "0x1",
                "0x0",
                "0x0",
                "0x1f40f0b0ad4322e13a35fae5400be2a2927a9b1ec7a1748158c1c456c49e56",
                "0x0",
                "0x5562b3e932b4d139366854d5a2e578382e6a3b6572ac9943d55e7efbe43d00",
                "0x2",
                "0x0",
                "0x1",
                "0x0",
                "0x1"
            ]
        },
        "block_id": "latest"
    },
    "id": 1
}'
```
returns:
```json
{
  "jsonrpc": "2.0",
  "result": [
    "0x1",
    "0x0",
    "0x0"
  ],
  "id": 1
}
```
To interpret as: `[ Result::Ok([]) ]`

A call to multicalls.raw_aggregate( [ dummy_contract.assert_eq(1, 0) ] ) failing (unexpected):
```json
curl --location 'https://free-rpc.nethermind.io/sepolia-juno/' \
--header 'Content-Type: application/json' \
--data '{
    "jsonrpc": "2.0",
    "method": "starknet_call",
    "params": {
        "request": {
            "contract_address": "0x07a9013697371ce40d0306b4c810c6a4db9bfda119dd9ae1e8701c8e288d734b",
            "entry_point_selector": "0x253d5ae28eea16e97fafcf8c5b2143ef58147be2687b9f15d9ec97992365fb8",
            "calldata": [
                "0x1",
                "0x0",
                "0x0",
                "0x1f40f0b0ad4322e13a35fae5400be2a2927a9b1ec7a1748158c1c456c49e56",
                "0x0",
                "0x5562b3e932b4d139366854d5a2e578382e6a3b6572ac9943d55e7efbe43d00",
                "0x2",
                "0x0",
                "0x1",
                "0x0",
                "0x0"
            ]
        },
        "block_id": "latest"
    },
    "id": 1
}'
```
Output:
```json
{
  "jsonrpc": "2.0",
  "error": {
    "code": 40,
    "message": "Contract error",
    "data": {
      "revert_error": "Error in the called contract (0x07a9013697371ce40d0306b4c810c6a4db9bfda119dd9ae1e8701c8e288d734b):\nError at pc=0:1141:\nGot an exception while executing a hint: Execution failed. Failure reason: 0x7920646f6573206e6f7420657175616c2079 ('y does not equal y').\nCairo traceback (most recent call last):\nUnknown location (pc=0:125)\n\nError in the called contract (0x001f40f0b0ad4322e13a35fae5400be2a2927a9b1ec7a1748158c1c456c49e56):\nExecution failed. Failure reason: 0x7920646f6573206e6f7420657175616c2079 ('y does not equal y').\n"
    }
  },
  "id": 1
}
```
Expected:
```json
{
  "jsonrpc": "2.0",
  "result": [
    "0x1",
    "0x1",
    "0x1",
    "0x7920646f6573206e6f7420657175616c2079"
  ],
  "id": 1
}
```
To interpret as: `[ Result::Err([ 0x7920646f6573206e6f7420657175616c2079 ]) ]`
